NeosModLoader/Utility/EnumerableInjector.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NeosModLoader.Utility
+{
+    /// <summary>
+    /// Provides the ability to inject actions into the execution of an enumeration while transforming it.
+    /// </summary>
+    /// <typeparam name="TIn">The type of the original enumeration.</typeparam>
+    /// <typeparam name="TOut">The type of the transformed enumeration.</typeparam>
+    public class EnumerableInjector<TIn, TOut> : IEnumerable<TOut>
+    {
+        /// <summary>
+        /// Internal enumerator for iteration.
+        /// </summary>
+        private readonly IEnumerator<TIn> enumerator;
+
+        private Action postfix = nothing;
+        private Action<TIn, TOut, bool> postItem = nothing;
+        private Action prefix = nothing;
+        private Func<TIn, bool> preItem = yes;
+        private Func<TIn, TOut> transformItem;
+
+        /// <summary>
+        /// Gets called when the wrapped enumeration returned the last item.
+        /// </summary>
+        public Action Postfix
+        {
+            get => postfix;
+            set => postfix = value ?? throw new ArgumentNullException(nameof(value), "Postfix can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item, with the transformed item, and whether it was passed through.
+        /// First thing to be called after execution returns to the enumerator after a yield return.
+        /// </summary>
+        public Action<TIn, TOut, bool> PostItem
+        {
+            get => postItem;
+            set => postItem = value ?? throw new ArgumentNullException(nameof(value), "PostItem can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called before the enumeration returns the first item.
+        /// </summary>
+        public Action Prefix
+        {
+            get => prefix;
+            set => prefix = value ?? throw new ArgumentNullException(nameof(value), "Prefix can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item to determine whether it should be passed through.
+        /// </summary>
+        public Func<TIn, bool> PreItem
+        {
+            get => preItem;
+            set => preItem = value ?? throw new ArgumentNullException(nameof(value), "PreItem can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item to transform it, even if it won't be passed through.
+        /// </summary>
+        public Func<TIn, TOut> TransformItem
+        {
+            get => transformItem;
+            set => transformItem = value ?? throw new ArgumentNullException(nameof(value), "TransforItem can't be null!");
+        }
```

Comment:
Is there a reason why the fields themselves aren't just public without the extra getters&seters?

Reply:
So that I can throw a null exception when someone attempts to set that?  
Also future expandibility n stuff

Reply:
We have `nullable` already enabled though, so like, it should complain if someone tries to shove a possibly null value in it?

Reply:
Does that affect places where nullable isn't enabled? 🤔